airbyte-cdk/python/airbyte_cdk/sources/streams/abstract_stream.py

```diff
@@ -0,0 +1,213 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import logging
+from abc import ABC, abstractmethod
+from functools import lru_cache
+from typing import Any, Iterable, List, Mapping, Optional, Tuple, Union
+
+import requests
+from airbyte_cdk.models import AirbyteMessage, AirbyteStream, SyncMode
+from airbyte_cdk.models import Type as MessageType
+from airbyte_cdk.sources.source import Source
+from airbyte_cdk.sources.utils import casing
+from airbyte_cdk.sources.utils.schema_helpers import InternalConfig
+from airbyte_cdk.sources.utils.slice_logger import SliceLogger
+from airbyte_cdk.sources.utils.transform import TransformConfig, TypeTransformer
+from airbyte_cdk.sources.utils.types import JsonType, StreamData
+
+
+class AbstractStream(ABC):
+    @abstractmethod
+    def read(
+        self,
+        cursor_field: Optional[List[str]],
+        logger: logging.Logger,
+        slice_logger: SliceLogger,
+        internal_config: InternalConfig = InternalConfig(),
+    ) -> Iterable[StreamData]:
```

Comment:
I don't like that this is returning a union of multiple things which requires the users to check what type of return this is (hence the `AbstractStream.is_record`. However, I don't have an easy solution to keep backward compatibility... I'm open to brainstorm on this